staging/src/k8s.io/api/coordination/v1beta1/types.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Lease defines a lease concept.
+type Lease struct {
+	metav1.TypeMeta `json:",inline"`
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the Lease.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec LeaseSpec `json:"spec,omitempty"`
+}
+
+// LeaseSpec is a specification of a Lease.
+type LeaseSpec struct {
+	// holderIdentity contains the identity of the holder of a current lease.
+	// +optional
+	HolderIdentity string `json:"holderIdentity"`
+	// leaseDurationSeconds is a duration that candidates for a lease need
+	// to wait to force acquire it. This is measure against time of last
+	// observed RenewTime.
+	// +optional
+	LeaseDurationSeconds int32 `json:"leaseDurationSeconds"`
+	// acquireTime is a time when the current lease was acquired.
+	// +optional
```

Comment:
I'm convinced that we need to adopt the proto2 convention of making every field optional to the decoding/encoding mechanism (which is why that notion was removed from proto3). I don't want client-side encodings with some fields absent (e.g., for patches) to have cruft in them (similar to unset creationTimestamp and empty status), and I don't want decoding to generate errors that should happen during schema validation.

While AcquireTime and LeaseTransitions really should be optional, I think we have to treat all of these fields as optional. They are all specified as optional in the comments, but they also need to be pointers and omitempty currently, unless something has changed.
https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#optional-vs-required

Hopefully we'll eventually build an IDL that would take care of such details.

cc @apelisse

Reply:
Re "omitempty" - added.

About making all those fields pointers - I can change that, but I would like to understand it better before.
Before creating that file I spent some time looking into API and it seems to me that this convention is a bit "dead" to me.
There is huge number of fields that has "+optional" comment but aren't pointers.
The appear in pretty much all api groups, e.g.:
- apps group: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/apps/v1/types.go#L226
[and many other examples in this file]
- core group: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L477
[and many other examples in this file]
- storage group: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/storage/v1beta1/types.go#L203
[and couple other examples in this file]
- ...

So we don't seem to have any consistency here anyway.
So I guess my question is: 
1. do we want to leave the type roughly the same as it already exists in annotation (without pointers)
or
2. do we want to change fields (all of them?) to pointers to apply convention (that doesn't seem to be applied in many places anyway)
or
3.  maybe mark only acquireTime and LeaseTransitions as optional?

@bgrant0607 - I initially decided to proceed with (1), but I would like to hear your guidance here.

Reply:
@lavalamp @erictune Could you please make a decision or delegate it to someone else in API machinery? I do expect these fields to be patched by clients. I don't really expect them to be declaratively applied, exported, or otherwise manipulated as configuration data.

Reply:
Yeah, the pointer is mostly useful to distinguish unset and set. If your field has a "non-valid" zero-value (for example, Time{} is not "valid" value, it's usually used as a sentinel value), then the pointer is not required.

I think in general we should keep it simple and just have the pointer all the time.

Reply:
@apelisse - so to clarify that I understand correctly - you suggest changing all fields to pointers, right?

Reply:
> you suggest changing all fields to pointers, right?

All optional field should be `omitempty` and pointers yes.